Add support for system_datastores_auth_mode in azurerm_machine_learning_workspace

bensincs opened the pull request:
<!--  All Submissions -->

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR adds support for the `system_datastores_auth_mode` property to the `azurerm_machine_learning_workspace` resource and data source. This property allows users to configure the authentication mode for system datastores in Azure Machine Learning workspaces.

The `system_datastores_auth_mode` field supports three values:
- `AccessKey` - Use access key authentication
- `Identity` - Use managed identity authentication  
- `UserDelegationSAS` - Use user delegation SAS authentication

This property is optional and can be set during workspace creation and updated after creation.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: "`azurerm_machine_learning_workspace` - support for the `system_datastores_auth_mode` property"

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

The changes include:
- Added the property to existing `complete` and `completeUpdated` tests with different values to test creation and updates
- Added validation for the data source test to verify the property is correctly exposed
- All tests validate that the property accepts the correct enum values and rejects invalid ones

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_machine_learning_workspace` - support for the `system_datastores_auth_mode` property [GH-22150]

<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
<!-- Add issue number if applicable -->
Issue: #22150 

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider. Since this is an optional property that doesn't affect existing configurations, rollback impact is minimal.

## Changes to Security Controls

This change adds support for configuring authentication modes for system datastores, which can enhance security by allowing users to choose more secure authentication methods like managed identity instead of access keys. No breaking changes to existing security controls are introduced.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

bensincs committed:
azurerm_machine_learning_workspace - support for the system_datastores_auth_mode property

This change adds support for configuring the authentication mode for system datastores in Azure Machine Learning workspaces.

The system_datastores_auth_mode property supports three authentication modes:
- AccessKey - Use access key authentication
- Identity - Use managed identity authentication
- UserDelegationSAS - Use user delegation SAS authentication

This property is optional and can be set during workspace creation and updated after creation.

Changes include:
- Added schema field with validation for the three supported values
- Implemented create, update, and read operations for the property
- Added the property to both resource and data source
- Updated tests to cover the new functionality
- Added documentation for both resource and data source

Co-authored-by: Vrushali Malushte <Vrushali.Malushte@microsoft.com>

magodo reviewed (changes requested):
Thanks for this PR!
I have left some additional mostly minor comments that once addressed this should be good to merge 👍

bensincs committed:
Apply suggestion from @magodo

Co-authored-by: magodo <wztdyl@sina.com>

bensincs committed:
Apply suggestion from @magodo

Co-authored-by: magodo <wztdyl@sina.com>

bensincs committed:
Merge remote-tracking branch 'origin' into ben

bensincs committed:
azurerm_machine_learning_workspace - support for the system_datastores_auth_mode property

This change adds support for configuring the authentication mode for system datastores in Azure Machine Learning workspaces.

The system_datastores_auth_mode property supports three authentication modes:
- AccessKey - Use access key authentication
- Identity - Use managed identity authentication
- UserDelegationSAS - Use user delegation SAS authentication

This property is optional and can be set during workspace creation and updated after creation.

Changes include:
- Added schema field with validation for the three supported values
- Implemented create, update, and read operations for the property
- Added the property to both resource and data source
- Updated tests to cover the new functionality
- Added documentation for both resource and data source

Co-authored-by: Vrushali Malushte <Vrushali.Malushte@microsoft.com>

magodo commented:
@bensincs One small change needed for the naming, can we use the full name `authentication` in favor of the abbreviation `auth`?

jackofallops commented on internal/services/machinelearning/machine_learning_workspace_data_source.go:
```diff
@@ -72,5 +77,9 @@ func dataSourceMachineLearningWorkspaceRead(d *pluginsdk.ResourceData, meta inte
 	}
 
+	if resp.Model.Properties != nil {
+		d.Set("system_datastores_auth_mode", pointer.From(resp.Model.Properties.SystemDatastoresAuthMode))
```
The `pointer` package is missing from the imports in this file, so this will not compile.
Also, `Model` is a pointer and needs to be nil-checked before accessing. The usage on line 72 above is also a panic risk, so we should check it before there.

jackofallops reviewed (changes requested):
Hi @bensincs - Applying the previous suggestions has unfortunately lead to a non-compilable PR. If you can take a look at the comment below, I'll take another look.
